src/toxcore/dht/kbucket.rs

```diff
@@ -76,36 +65,88 @@ impl Distance for PublicKey {
     }
 }
 
+/// Anything that has `PublicKey`.
+pub trait HasPK {
+    /// `PublicKey`.
+    fn pk(&self) -> PublicKey;
+}
+
+impl HasPK for PackedNode {
+    fn pk(&self) -> PublicKey {
+        self.pk
+    }
+}
+
+/// Node that can be stored in a `Kbucket`.
+pub trait KbucketNode : Sized + HasPK {
+    /// The type of nodes that can be added to a `Kbucket`.
+    type NewNode: HasPK;
+
+    /// Check if the node can be updated with a new one.
+    fn is_outdated(&self, other: &Self::NewNode) -> bool;
+    /// Update the existing node with a new one.
+    fn update(&mut self, other: &Self::NewNode);
+    /// Check if the node can be evicted.
+    fn is_evictable(&self) -> bool;
+    /// Find the index of a node that should be evicted in case if `Kbucket` is
+    /// full. It must return `Some` if and only if nodes list contains at least
+    /// one evictable node.
+    fn eviction_index(nodes: &[Self]) -> Option<usize>;
+}
+
+impl KbucketNode for PackedNode {
+    type NewNode = PackedNode;
```

Comment:
Is there any impl KbucketNode with NewNode != PacketNode ?

Reply:
Not right now but will be for onion client. It will an implementation for another type `OnionNode`. The reason why I need it is that `OnionNode` will store additional info, like onion path id. And it will be not enough to have only `PackedNode` to construct such node.